README.md
<p align="center">
  <a href="https://powerledger.io">
    <img alt="powr" src="https://github.com/PowerLedger/powr/raw/master/logo.png" width="250" />
  </a>
</p>

# POWERLEDGER ENERGY BLOCKCHAIN

## IMPORTANT NOTICE - PLEASE READ BEFORE PROCEEDING

Use of this software and access to the Powerledger Blockchain is subject to the Powerledger Validator
Node Terms and Conditions set out in the Terms and Conditions file.

By downloading, installing and using the Powerledger Node Software and/or connecting to the Powerledger
Blockchain you agree to the Terms and Conditions set out in the Terms and Conditions.

All software is Copyright © 2023 Power Ledger Pty Ltd unless indicated otherwise.  Unauthorised use
prohibited. All Rights Reserved.

## powr validator node configuration

### Min Requirements for mainnet validator

```
* 12 Cores, 64 GB RAM
* ┬ 3 separate storage devices:
  ├*- 250 GB of storage for the operating system
  ├*- 1 TB of storage for the accounts folder
  └*- 4 TB of storage for the ledger folder (pruned nodes) 
* Rust 1.59 installed
```

### Download the project

```bash
# Git pull powr and powr-program-library repositories
# Both repos need to be downloaded in the same location
git clone https://github.com/powerledger/powr
git clone https://github.com/powerledger/powr-program-library
```

## Compile the code

```bash
cd powr
cargo clean
cd scripts
./cargo-install-all.sh /home/validator-admin/.local/share/powr
```

## Append to .bashrc

```bash
echo -e "\nexport PATH=\$PATH:/home/validator-admin/.local/share/powr/bin" >> ~/.bashrc
cd
. .bashrc
```

## Create identities

At least 2 keypairs are necessary, and IDENTITY keypair and a VOTE-ACOUNT keypair.

We suggest creating a third keypair that is going to function as the withdrawal authority
for the vote account.

```bash
# Keypair generation example: 

mkdir ~/keypairs
solana-keygen new --word-count 24 --no-bip39-passphrase -o ~/keypairs/id.json
solana-keygen new --word-count 24 --no-bip39-passphrase -o ~/keypairs/vote-account.json
solana-keygen new --word-count 24 --no-bip39-passphrase -o ~/keypairs/withdraw-account.json
chmod 400 ~/keypairs/*
```

## Create config file

Execute the following commands to create the config file:

```bash

```bash
[ ! -d "~/.config/solana/cli" ] && mkdir -p ~/.config/solana/cli
cat > ~/.config/solana/cli/config.yml << 'EOF'
---
json_rpc_url: "https://powr-api.mainnet.powerledger.io"
websocket_url: ""
keypair_path: $HOME/keypairs/id.json
address_labels:
  "11111111111111111111111111111111": SystemProgram
commitment: confirmed
EOF
```

If your IDENTITY keypair is located in a different folder, it can be changed using the
following command:

```bash
solana config set -k /PATH_TO_KEYPAIR/id.json
```

## Create vote account

Example usage:

```bash
solana create-vote-account /PATH_TO_KEYPAIR/vote-account.json /PATH_TO_KEYPAIR/id.json $(solana address -k /PATH_TO_KEYPAIR/withdraw-account.json) --commission <0-100>
```

If the commission needs to be updated:

```bash
solana vote-update-commission $(solana address -k /PATH_TO_KEYPAIR/vote-account.json) <0-100> /PATH_TO_KEYPAIR/withdraw-account.json
```

## Starting you validator node

IMPORTANT NOTE: You can modify the command to fit your requirements
but the list of --known-validator(s) must be provided.

### Full history node

```bash
#!/usr/bin/env bash

exec <PATH TO>/solana-validator \
--ledger <PATH TO>/ledger \
--accounts <PATH TO>/accounts \
--full-rpc-api \
--private-rpc \
--rpc-port 8899 \
--snapshot-interval-slots 200 \
--account-index program-id \
--account-index spl-token-mint \
--account-index spl-token-owner \
--identity <PATH TO>/keypairs/id.json \
--authorized-voter <PATH TO>/keypairs/id.json \
--vote-account <PATH TO>/keypairs/vote-account.json \
--expected-genesis-hash B1xJgyn5UyWw2itxwwctHL5fLxY9ZbnsDf7AG2WXbpGY \
--known-validator 8cDZLH2Ha3v56WLcpQkH7U211uokkb6YbZ3Sc6r6WrUa \
--known-validator 6YDqUUE3JBA9sVLNNiWYQR7moAPyNfDsBKPEvGcuEfQ2 \
--known-validator 8oATwTdKJqUvWMkUjFDBVW23HDF5DbVRskxWCypXwRRB \
--known-validator 6H69egwLDXs27r7tfUu5oxesbcYbi6YL8JW2D22jkBEi \
--wal-recovery-mode skip_any_corrupted_record \
--enable-rpc-transaction-history \
--enable-cpi-and-log-storage \
--entrypoint powr-entrypoint-1.mainnet.powerledger.io:8001 \
--entrypoint powr-entrypoint-2.mainnet.powerledger.io:8001 \
--dynamic-port-range 8000-9000 \
--log <PATH TO>/validator.log
```

### Pruned Node

```bash
#!/usr/bin/env bash

exec <PATH TO powr/target/release>/solana-validator \
--ledger <PATH TO>/ledger \
--accounts <PATH TO>/accounts \
--full-rpc-api \
--private-rpc \
--rpc-port 8899 \
--account-index program-id \
--account-index spl-token-mint \
--account-index spl-token-owner \
--identity <PATH TO>/keypairs/id.json \
--authorized-voter <PATH TO>/keypairs/id.json \
--vote-account <PATH TO>/keypairs/vote-account.json \
--expected-genesis-hash B1xJgyn5UyWw2itxwwctHL5fLxY9ZbnsDf7AG2WXbpGY \
--known-validator 8cDZLH2Ha3v56WLcpQkH7U211uokkb6YbZ3Sc6r6WrUa \
--known-validator 6YDqUUE3JBA9sVLNNiWYQR7moAPyNfDsBKPEvGcuEfQ2 \
--known-validator 8oATwTdKJqUvWMkUjFDBVW23HDF5DbVRskxWCypXwRRB \
--known-validator 6H69egwLDXs27r7tfUu5oxesbcYbi6YL8JW2D22jkBEi \
--wal-recovery-mode skip_any_corrupted_record \
--enable-rpc-transaction-history \
--entrypoint powr-entrypoint-1.mainnet.powerledger.io:8001 \
--entrypoint powr-entrypoint-2.mainnet.powerledger.io:8001 \
--dynamic-port-range 8000-9000 \
--limit-ledger-size \
--log <PATH TO>/validator.log
```

We HIGHLY recommend to manage the process using linux's `systemctl` command:

```bash
sudo su
cat > /etc/systemd/system/powr.service << 'EOF'
[Unit]
Description=Powerledger Blockchain Validator Service
Requires=network-online.target
After=network.target

[Service]
LimitNOFILE=1000000
Type=simple
User=<USER>
WorkingDirectory=/home/<USER>
ExecStart=<PATH TO SOLANA VALIDATOR SCRIPT>
Restart=always
RestartSec=1

[Install]
WantedBy=multi-user.target
EOF
exit
```

## System tunning

```bash
# Increase UDP buffers
sudo bash -c "cat >/etc/sysctl.d/20-solana-udp-buffers.conf << 'EOF'
# Increase UDP buffer size
net.core.rmem_default = 134217728
net.core.rmem_max = 134217728
net.core.wmem_default = 134217728
net.core.wmem_max = 134217728
EOF"
sudo sysctl -p /etc/sysctl.d/20-solana-udp-buffers.conf
```

```bash
# Increased memory mapped files limit
sudo bash -c "cat >/etc/sysctl.d/20-solana-mmaps.conf << 'EOF'
# Increase memory mapped files limit
vm.max_map_count = 1000000
EOF"
sudo sysctl -p /etc/sysctl.d/20-solana-mmaps.conf
```

```bash
# Add the following line to the [Service] section of your systemd service file, if you use one
LimitNOFILE=1000000
```

```bash
# Increasing max opened files limit
sudo bash -c "cat >/etc/security/limits.d/90-solana-nofiles.conf << 'EOF'
# Increase process file descriptor count limit
* - nofile 1000000
EOF"
```

```bash
# Close all open sessions (log out then, in again)
# check new limit
ulimit -Sn
```

## Create logrotate configuration

```bash
sudo su
cat > /etc/logrotate.d/<NAME OF THE systemctl SERVICE>.conf << 'EOF'
/PATH_TO_LOGFILE/validator.log {
    rotate 7
    daily
    missingok
    postrotate
        systemctl kill -s USR1 <NAME OF THE systemctl SERVICE>.service
    endscript
}
EOF
systemctl restart logrotate.service
exit
```

## Withdrawing tokens from vote-account

```bash
solana withdraw-from-vote-account \
$(solana-keygen pubkey /PATH_TO_KEYPAIR/vote-account.json) \
<RECIPIENT ADDRESS> \
<AMOUNT> \
--withdraw-authority /PATH_TO_KEYPAIR/withdraw-account.json
```

## Publish additional information about the validator

```bash
solana validator-info publish -k /PATH_TO_KEYPAIR/id.json -w "<URL OF THE VALIDATOR>" "<VALIDATOR NAME>"
```

## Getting published validators metadata

```bash
solana validator-info get
```


RELEASE_NOTES.md
# RELEASE NOTES

## v1.13.7

* increases retransmit-stage deduper capacity and reset-cycle (backport of #30758) (#30857)
* Generalizes deduper to work with any hashable type (backport of #30753) (#30854)
* Adds metrics tracking deduper saturations (backport of #30779) (#30847)
* Panic when shred index exceeds the max per slot (v1.13) (#30623)
* Removes the false_positive_rate field from the Deduper (backport of #30788) (#30839)
* Increases shred-fetch-stage deduper capacity and reset-cycle (backport of #30690) (#30835)
* Dedups packets using an atomic bloom filter (backport of #30726) (#30826)
* ledger-tool: Retain all snapshots by default (#30768)
* Stops nodes from broadcasting slots twice (backport of #30681) (#30684)
* Restrict QUIC to use single self signed client cert (backport #29681) (#29683)

## v1.13.6

* Change SlotMeta is_connected bool to bitflags (backport #29001) (#29020)
* use signed repair request variants (backport #28283) (#28988)
* enforces hash domain for ping-pong protocol (backport #28433) (#28961)
* Fix client get_program_accounts_with_config calls with context (backport #28772) (#28846)
* signed repair request test fixes/cleanup (backport #28691) (#28824)
* bigtable: add timeout to token refresh (backport #28728) (#28807)

## v1.13.5

* Validators on 1.13.4 or 1.13.5 will no longer support the UDP protocol for transaction ingress by default. If you are currently using --tpu-disable-quic, make sure to remove it.
* 1.13.5 contains important QUIC client performance improvements - upgrading RPC nodes is strongly recommended.
* If you are building your own binaries make sure to use the correct Rust version as specified in ci/rust-version.sh (1.59 for v1.13). More recent Rust versions contain backwards-incompatible changes which break the QUIC library that Solana uses. Solana comes with a wrapper script ( ./cargo ) which automatically uses the correct version.
* Commits
* MultiIteratorScanner - improve banking stage performance with high contention
* Delete files older than the lowest_cleanup_slot in LedgerCleanupService::cleanup_ledger (backport #26651) (#28721)
* This commit will significantly improve the ledger's ability to reclaim disk space, and thus, remain closer to expected size in worst case.
* Backport timeout async send tasks to v1.13 (#28750)
* clean_dead_slots_from_accounts_index unrefs correctly (backport #2746… (#28582)
* Limit async tasks spawn to runtime for quic transaction send (backport #28452) (#28603)
* Backport of #28599 to v1.13 (#28602)
* stats for staked/unstaked repair requests (backport #28215) (#28598)
* make ping cache rate limit delay configurable (backport #27955) (#28595)
* ledger-tool: remove inefficient base58 encoding options (backport #28488) (#28494)

## v1.13.4

* This release contains changes to disable the validator's UDP TPU port by default. It can be re-enabled by passing the --tpu-enable-udp flag.

### New v1.13 Commits

* bootstrap: Require known validators to have all snapshot hashes (backport #28284) (#28476)
* Added option to turn on UDP for TPU transaction and make UDP based TPU off by default (backport #27462) (#27658)

## v1.13.3

* Fixed the local-cluster test case test_optimistic_confirmation_violation_detection (backport #27580) (#27649)

## v1.13.2

* Nodes with 0% commission are now included in the getBlock RPC method rewards response field. (backport #28001) (#28010)
* Update CI error message when the version number needs to be bumped (backport #27997) (#28020)
* docs: fix wrong args in 'solana program set-buffer-authority' (backport #27817) (#28045)
* allow unsigned repair requests (backport #27910) (#28026)
* count unsigned repair requests (backport #27953) (#28067)
* Fixed the local-cluster test case test_optimistic_confirmation_violation_detection (backport #27580) (#27649)
* counts gossip packets received before excess packets are dropped (backport #28086) (#28105)
* Allow validators to reset to the slot which matches their last voted slot (backport #28172) (#28186)

## v1.13.1 Pre-release

* Fix local-cluster for QUIC more (backport #27096) (#27103)
* Fix local cluster tests for QUIC usage (backport #27071) (#27077)
* Use batch send in bench-tps to send transactions (backport #27527) (#27560)
* Fix quic staked chunking (backport #27402) (#27408)
* Enable QUIC client by default. Add arg to disable QUIC client. Take 2 (#26927)
* Fix quic client on TestValidator, alternative (backport #27046) (#27053)
* Compute maximum parallel QUIC streams using client stake (#26802)
* Fix transaction chunking on QUIC batch send (#26642)
* Merge pull request from GHSA-x236-qc46-7v8j
* Remove some unchecked math from the quic chunk handling (#26408)
* Fix new 1.62 clippy complaint
* Replace unwraps with expects in quic-client to aid debugging (#26283)
* tpu-client: Refactor to prep for async client (#25432)


cli-config/src/config.rs
// Wallet settings that can be configured for long-term use
use {
    serde_derive::{Deserialize, Serialize},
    std::{collections::HashMap, io, path::Path},
    url::Url,
};

lazy_static! {
    /// The default path to the CLI configuration file.
    ///
    /// This is a [lazy_static] of `Option<String>`, the value of which is
    ///
    /// > `~/.config/solana/cli/config.yml`
    ///
    /// It will only be `None` if it is unable to identify the user's home
    /// directory, which should not happen under typical OS environments.
    ///
    /// [lazy_static]: https://docs.rs/lazy_static
    pub static ref CONFIG_FILE: Option<String> = {
        dirs_next::home_dir().map(|mut path| {
            path.extend(&[".config", "solana", "cli", "config.yml"]);
            path.to_str().unwrap().to_string()
        })
    };
}

/// The Solana CLI configuration.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct Config {
    /// The RPC address of a Solana validator node.
    ///
    /// Typical values for mainnet, devnet, and testnet are [described in the
    /// Solana documentation][rpcdocs].
    ///
    /// For local testing, the typical value is `http://localhost:8899`.
    ///
    /// [rpcdocs]: https://docs.solana.com/cluster/rpc-endpoints
    pub json_rpc_url: String,
    /// The address to connect to for receiving event notifications.
    ///
    /// If it is an empty string then the correct value will be derived
    /// from `json_rpc_url`.
    ///
    /// The default value is the empty string.
    pub websocket_url: String,
    /// The default signing source, which may be a keypair file, but may also
    /// represent several other types of signers, as described in the
    /// documentation for `solana_clap_utils::keypair::signer_from_path`.
    /// Because it represents sources other than a simple path, the name
    /// `keypair_path` is misleading, and exists for backwards compatibility
    /// reasons.
    ///
    /// The signing source can be loaded with either the `signer_from_path`
    /// function, or with `solana_clap_utils::keypair::DefaultSigner`.
    pub keypair_path: String,
    /// A mapping from Solana addresses to human-readable names.
    ///
    /// By default the only value in this map is the system program.
    #[serde(default)]
    pub address_labels: HashMap<String, String>,
    /// The default commitment level.
    ///
    /// By default the value is "confirmed", as defined by
    /// `solana_sdk::commitment_config::CommitmentLevel::Confirmed`.
    #[serde(default)]
    pub commitment: String,
}

impl Default for Config {
    fn default() -> Self {
        let keypair_path = {
            let mut keypair_path = dirs_next::home_dir().expect("home directory");
            keypair_path.extend(&[".config", "solana", "id.json"]);
            keypair_path.to_str().unwrap().to_string()
        };
        let json_rpc_url = "https://plb-api.mainnet.powerledger.io".to_string();

        // Empty websocket_url string indicates the client should
        // `Config::compute_websocket_url(&json_rpc_url)`
        let websocket_url = "".to_string();

        let mut address_labels = HashMap::new();
        address_labels.insert(
            "11111111111111111111111111111111".to_string(),
            "System Program".to_string(),
        );

        let commitment = "confirmed".to_string();

        Self {
            json_rpc_url,
            websocket_url,
            keypair_path,
            address_labels,
            commitment,
        }
    }
}

impl Config {
    /// Load a configuration from file.
    ///
    /// # Errors
    ///
    /// This function may return typical file I/O errors.
    pub fn load(config_file: &str) -> Result<Self, io::Error> {
        crate::load_config_file(config_file)
    }

    /// Save a configuration to file.
    ///
    /// If the file's directory does not exist, it will be created. If the file
    /// already exists, it will be overwritten.
    ///
    /// # Errors
    ///
    /// This function may return typical file I/O errors.
    pub fn save(&self, config_file: &str) -> Result<(), io::Error> {
        crate::save_config_file(self, config_file)
    }

    /// Compute the websocket URL from the RPC URL.
    ///
    /// The address is created from the RPC URL by:
    ///
    /// - adding 1 to the port number,
    /// - using the "wss" scheme if the RPC URL has an "https" scheme, or the
    ///   "ws" scheme if the RPC URL has an "http" scheme.
    ///
    /// If `json_rpc_url` cannot be parsed as a URL then this function returns
    /// the empty string.
    pub fn compute_websocket_url(json_rpc_url: &str) -> String {
        let json_rpc_url: Option<Url> = json_rpc_url.parse().ok();
        if json_rpc_url.is_none() {
            return "".to_string();
        }
        let json_rpc_url = json_rpc_url.unwrap();
        let is_secure = json_rpc_url.scheme().to_ascii_lowercase() == "https";
        let mut ws_url = json_rpc_url.clone();
        ws_url
            .set_scheme(if is_secure { "wss" } else { "ws" })
            .expect("unable to set scheme");
        if let Some(port) = json_rpc_url.port() {
            let port = port.checked_add(1).expect("port out of range");
            ws_url.set_port(Some(port)).expect("unable to set port");
        }
        ws_url.to_string()
    }

    /// Load a map of address/name pairs from a YAML file at the given path and
    /// insert them into the configuration.
    pub fn import_address_labels<P>(&mut self, filename: P) -> Result<(), io::Error>
    where
        P: AsRef<Path>,
    {
        let imports: HashMap<String, String> = crate::load_config_file(filename)?;
        for (address, label) in imports.into_iter() {
            self.address_labels.insert(address, label);
        }
        Ok(())
    }

    /// Save the map of address/name pairs contained in the configuration to a
    /// YAML file at the given path.
    pub fn export_address_labels<P>(&self, filename: P) -> Result<(), io::Error>
    where
        P: AsRef<Path>,
    {
        crate::save_config_file(&self.address_labels, filename)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn compute_websocket_url() {
        assert_eq!(
            Config::compute_websocket_url("http://api.devnet.solana.com"),
            "ws://api.devnet.solana.com/".to_string()
        );

        assert_eq!(
            Config::compute_websocket_url("https://api.devnet.solana.com"),
            "wss://api.devnet.solana.com/".to_string()
        );

        assert_eq!(
            Config::compute_websocket_url("http://example.com:8899"),
            "ws://example.com:8900/".to_string()
        );
        assert_eq!(
            Config::compute_websocket_url("https://example.com:1234"),
            "wss://example.com:1235/".to_string()
        );

        assert_eq!(Config::compute_websocket_url("garbage"), String::new());
    }
}
